### PR TITLE
Support Reusable Request Bodies and Improve Payload Handling

### LINKED ISSUE #5530: Redirect on upload from file causes IO error from file being closed
🐞 **Describe the bug**
As seen from #1907, `aiohttp` takes ownership of passed files and closes them after writing. However, in the case of a 302 redirect, where I assume that `aiohttp` is re-performing the request, it attempts to re-read the data from the closed file.

💡 **To Reproduce**

Client code:
```python
from aiohttp import ClientSession
with open(abs_storage_path(rel_path), "rb") as f:
    async with ClientSession() as session:
        async with session.put("http://localhost:5000/foo", data=f) as r:
            print("Response:", r.status)
```

Server code:
```python
from flask import Flask, request, redirect

app = Flask(__name__)

@app.route("/foo", methods=['PUT'])
def redirct_route():
    return redirect("http://localhost:5000/bar", code=302)

@app.route("/bar", methods=['PUT'])
def put_route():
    print('Uploading data of size:', len(request.data))
    return 'success', 200

app.run(debug=True)
```

💡 **Expected behavior**
Successful PUT upload (or to not follow redirects and return a 302 response).

📋 **Logs/tracebacks**
```
    async with session.put("http://localhost:5000/foo", data=f) as r:
               │       │                                     └ <_io.BufferedReader name='../data/foo.dat'>
               │       └ <function ClientSession.put at 0x7faa1294de50>
               └ <aiohttp.client.ClientSession object at 0x7faa101d9a00>

  File ".../aiohttp/client.py", line 1117, in __aenter__
    self._resp = await self._coro
    │    │             │    └ <member '_coro' of '_BaseRequestContextManager' objects>
    │    │             └ <aiohttp.client._RequestContextManager object at 0x7faa101e7bc0>
    │    └ <member '_resp' of '_BaseRequestContextManager' objects>
    └ <aiohttp.client._RequestContextManager object at 0x7faa101e7bc0>
  File ".../aiohttp/client.py", line 492, in _request
    req = self._request_class(
          │    └ <class 'aiohttp.client_reqrep.ClientRequest'>
          └ <aiohttp.client.ClientSession object at 0x7faa101d9a00>
  File ".../aiohttp/client_reqrep.py", line 313, in __init__
    self.update_body_from_data(data)
    │    │                     └ <_io.BufferedReader name='../data/foo.dat'>
    │    └ <function ClientRequest.update_body_from_data at 0x7faa12c5e0d0>
    └ <aiohttp.client_reqrep.ClientRequest object at 0x7faa1015e580>
  File ".../aiohttp/client_reqrep.py", line 519, in update_body_from_data
    size = body.size
           │    └ <property object at 0x7faa12c8ecc0>
           └ <aiohttp.payload.BufferedReaderPayload object at 0x7faa1015e730>
  File ".../aiohttp/payload.py", line 362, in size
    return os.fstat(self._value.fileno()).st_size - self._value.tell()
           │  │     │    │      │                   │    │      └ <method 'tell' of '_io.BufferedReader' objects>
           │  │     │    │      │                   │    └ <_io.BufferedReader name='../data/foo.dat'>
           │  │     │    │      │                   └ <aiohttp.payload.BufferedReaderPayload object at 0x7faa1015e730>
           │  │     │    │      └ <method 'fileno' of '_io.BufferedReader' objects>
           │  │     │    └ <_io.BufferedReader name='../data/foo.dat'>
           │  │     └ <aiohttp.payload.BufferedReaderPayload object at 0x7faa1015e730>
           │  └ <built-in function fstat>
           └ <module 'os' from '/usr/lib/python3.9/os.py'>

ValueError: I/O operation on closed file
```

📋 **Your version of the Python**

```console
$ python --version
Python 3.9.1
```

📋 **Your version of the aiohttp/yarl/multidict distributions**

```console
$ python -m pip show aiohttp
Name: aiohttp
Version: 3.7.4
Summary: Async http client/server framework (asyncio)
Home-page: https://github.com/aio-libs/aiohttp
Author: Nikolay Kim
Author-email: fafhrd91@gmail.com
License: Apache 2
Location: .../python3.9/site-packages
Requires: yarl, attrs, typing-extensions, chardet, async-timeout, multidict
```
```console
$ python -m pip show multidict
Name: multidict
Version: 5.1.0
Summary: multidict implementation
Home-page: https://github.com/aio-libs/multidict
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: .../python3.9/site-packages
Requires: 
Required-by: yarl, async-asgi-testclient, aiohttp
```
```console
$ python -m pip show yarl
Name: yarl
Version: 1.6.3
Summary: Yet another URL library
Home-page: https://github.com/aio-libs/yarl/
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: .../python3.9/site-packages
Requires: idna, multidict
Required-by: aiohttp
```

📋 **Additional context**
Relates to the **client** aspect of aiohttp.

### LINKED ISSUE #5577: Incorrect reuse of FormData on certain redirects
🐞 **Describe the bug**

If a POST request results in a redirect response with response code 307 or 308, `aiohttp.client` may attempt to construct another POST request with the redirected URL using the same `data` as the first request. If the value passed as the `data` argument is an instance of `aiohttp.FormData`, this will fail with a somewhat cryptic exception: ["RuntimeError: Form data has been processed already"](https://github.com/aio-libs/aiohttp/blob/0c36ab6350f8c82ac467d23183eb42ead515754d/aiohttp/formdata.py#L132). 

💡 **To Reproduce**

1. Send a POST request with FormData to a server that will respond with code 307 or 308

💡 **Expected behavior**

Instead of reconstructing the body for each request, perhaps the request body could be reused. This would also resolve #5530, which appears to have a similar root cause.

📋 **Logs/tracebacks**

```python-traceback (paste your traceback in the next line)
  File "--", line --, in --
    async with session.post(url, data=data) as response:
  File "/usr/local/lib/python3.8/dist-packages/aiohttp/client.py", line 1118, in __aenter__
    self._resp = await self._coro
  File "/usr/local/lib/python3.8/dist-packages/aiohttp/client.py", line 493, in _request
    req = self._request_class(
  File "/usr/local/lib/python3.8/dist-packages/aiohttp/client_reqrep.py", line 313, in __init__
    self.update_body_from_data(data)
  File "/usr/local/lib/python3.8/dist-packages/aiohttp/client_reqrep.py", line 507, in update_body_from_data
    body = body()
  File "/usr/local/lib/python3.8/dist-packages/aiohttp/formdata.py", line 169, in __call__
    return self._gen_form_data()
  File "/usr/local/lib/python3.8/dist-packages/aiohttp/formdata.py", line 133, in _gen_form_data
    raise RuntimeError("Form data has been processed already")
RuntimeError: Form data has been processed already
```

📋 **Your version of the Python**

```console
$ python --version
Python 3.8.5
```

📋 **Your version of the aiohttp/yarl/multidict distributions**

```console
$ python -m pip show aiohttp
Name: aiohttp
Version: 3.7.4.post0
Summary: Async http client/server framework (asyncio)
Home-page: https://github.com/aio-libs/aiohttp
Author: Nikolay Kim
Author-email: fafhrd91@gmail.com
License: Apache 2
Location: /usr/local/lib/python3.8/dist-packages
Requires: attrs, chardet, yarl, multidict, typing-extensions, async-timeout
```
```console
$ python -m pip show multidict
Version: 5.1.0
Summary: multidict implementation
Home-page: https://github.com/aio-libs/multidict
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: /usr/local/lib/python3.8/dist-packages
Requires: 
Required-by: yarl, aiohttp
```
```console
$ python -m pip show yarl
Name: yarl
Version: 1.6.3
Summary: Yet another URL library
Home-page: https://github.com/aio-libs/yarl/
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: /usr/local/lib/python3.8/dist-packages
Requires: idna, multidict
Required-by: aiohttp
```

📋 **Additional context**

This is a client issue.

For anyone else who has encountered this issue: Skipping the redirect by sending the original request to the redirected URL may be an adequate workaround at least in the case of a permanent redirect.

### PR DESCRIPTION
**Pull request by bdraco**
## Overview
This PR enhances aiohttp's payload handling system to support reusable request bodies, addressing a common pain point where payloads could only be used once. The changes enable retry mechanisms and middleware to work with request bodies without consuming them permanently.

Fixes #5577 - Digest authentication now works with request bodies (previously failed with "payload consumed" errors)
Fixes #5530 - Request bodies can now be reused in retry scenarios without "attempt to read body twice" errors
Replaces and closes #9201 - Thanks @GLGDLY

### 1. Reusable Payload Support
- Added `consumed` and `autoclose` properties to track payload state
- Implemented `as_bytes()` method for safe payload content retrieval
- Added caching mechanism for `AsyncIterablePayload` to enable reuse
- File-based payloads now remember their start position for re-reading

### 2. Enhanced Payload Classes
- **BytesPayload**: Marked as `autoclose=True` since bytes are immutable and reusable
- **BytesIOPayload**: Optimized `size` property using `cached_property` 
- **IOBasePayload**: Added position tracking and restoration for file handles
- **AsyncIterablePayload**: Implemented chunk caching to allow multiple reads

### 3. ClientRequest Improvements
- Added `update_body()` method for middleware use to safely update request bodies
- Made sure the `.body` setter cleans up all current payload types, and documented that using the setter is the wrong way to do this as we want consumers of the API to call the async `update_body()` method so we can safely clean up any type of new payloads in the future. A resource warning will be thrown if they manually set the body using the setter instead of calling `update_body()`, as this is likely to leak in the future.
- Improved handling of empty bodies and proper cleanup
- Better separation between body preparation and sending

### 4. Middleware Enhancements
- Updated digest authentication middleware to use new reusable payload features
- Added support for middleware to safely inspect and modify request bodies
- Improved error handling for consumed payloads

### 5. Resource Management
- Implemented proper `close()` methods for all payload types
- Fixed resource leaks by scheduling file closes without awaiting
- Better handling of payload cleanup during redirects and errors

## Benefits
- Enables retry middleware to resend requests with bodies
- Allows digest authentication to work with request bodies
- Supports middleware that needs to inspect request content
- Prevents "payload already consumed" errors in common scenarios
- Improves resource cleanup and prevents file descriptor leaks

## Testing
- Added comprehensive tests for payload reusability
- Tests for middleware interaction with request bodies
- Benchmarks to ensure no performance regression
- Edge case testing for file position restoration and async iterables

## Breaking Changes
None - all changes are backward compatible. Existing code will continue to work as before.

## Documentation
- Updated client reference documentation
- Added examples of using reusable payloads
- Documented best practices for middleware authors

### CODE REVIEW

File: aiohttp/client_reqrep.py [hunk 1]
```diff
@@ -97,6 +97,8 @@
 
 _CONTAINS_CONTROL_CHAR_RE = re.compile(r"[^-!#$%&'*+.^_`|~0-9a-zA-Z]")
 
+BodyType = Union[payload.Payload, bytes, bytearray, memoryview]
```
Comment (Dreamsorcerer) [inline aiohttp/client_reqrep.py]: These other types will actually never happen (except `b""`). Every type is thrown into the registry or loaded from FormData:
https://github.com/aio-libs/aiohttp/blob/b9b19f659170e8cfd7a26f1d50a56a6c0b8ebd97/aiohttp/client_reqrep.py#L581

I'd like to avoid adding any of these aliases, as I'm currently trying to reduce this code to _only_ `Payload` and remove all the extra code that would never be used.

File: aiohttp/client_reqrep.py [hunk 2]
```diff
@@ -560,7 +587,9 @@ def update_body_from_data(self, body: Any) -> None:
                 )
             body = FormData(body, boundary=boundary)()
 
-        self.body = body
+        self._body = body
+        # At this point, body is always a Payload (either from PAYLOAD_REGISTRY or FormData())
+        self._body_is_payload = True
```
Comment (Dreamsorcerer) [inline aiohttp/client_reqrep.py]: Likewise, I'd prefer to stick to an isinstance() check, as I'm already removing any code that is not related to Payload. So, this will just be more stuff to revert.

Comment (bdraco) [reply aiohttp/client_reqrep.py]: isinstance is a bit expensive considering how many possible inheritances we have here we will now need to do it a lot more places now

Comment (Dreamsorcerer) [reply aiohttp/client_reqrep.py]: Could add the optimisation in 3.x branch only, if my changes can't be backported?

Comment (bdraco) [reply aiohttp/client_reqrep.py]: I came up with a way to avoid the isinstance checks

File: aiohttp/client_reqrep.py (hunk 1 above)
Comment (bdraco) [reply aiohttp/client_reqrep.py]: removed

File: aiohttp/client_reqrep.py [hunk 3]
```diff
@@ -206,7 +220,7 @@ class ClientRequest:
     }
 
     # Type of body depends on PAYLOAD_REGISTRY, which is dynamic.
-    body: Any = b""
+    _body: Union[bytes, payload.Payload] = b""
```
Comment (bdraco) [inline aiohttp/client_reqrep.py]: We should initialize to None as it makes it so much more sane

Comment (Dreamsorcerer) [reply aiohttp/client_reqrep.py]: The refactor initialises to an empty BytesPayload..

File: aiohttp/payload.py [hunk 4]
```diff
@@ -827,10 +966,27 @@ async def write_with_length(
...
+        """
+        if self._iter is None:
+            return b""
+        chunks: List[bytes] = []
+        async for chunk in self._iter:
+            chunks.append(chunk)
+        # Iterator is exhausted after consuming
+        self._iter = None
+        self._consumed = True  # Iterator has been consumed
+        return b"".join(chunks)
```
Comment (Dreamsorcerer) [inline aiohttp/payload.py:1069]: If we're going to load the entire thing into memory anyway, it could be used to make it re-consumable. Otherwise, there's probably no point to implementing this method.

File: tests/test_client_request.py [hunk 5]
```diff
@@ -1635,7 +1635,17 @@ async def test_write_bytes_with_iterable_content_length_limit(
...
+    # Convert list to async generator if needed
+    if isinstance(data, list):
+
+        async def gen() -> AsyncIterator[bytes]:
+            for chunk in data:
+                yield chunk
+
+        req.body = gen()
+    else:
+        req.body = data
```
Comment (Dreamsorcerer) [inline tests/test_client_request.py]: These kind of tests are redundant, given that no actual production code will ever execute this. It'd be more useful to pass `data` in to the ClientRequest construction above, if there is still something relevant to test.

But, there's no point having a test that verifies that write_bytes() works when body is a list etc. because that can never happen in actual use.

File: aiohttp/payload.py (hunk 4 above)
Comment (bdraco) [reply aiohttp/payload.py:1069]: abc906b5d

needs tests

File: tests/test_client_request.py (hunk 5 above)
Comment (Dreamsorcerer) [reply tests/test_client_request.py]: I guess you've put a setter on it, though I wonder if we shouldn't discourage that..

Comment (bdraco) [reply tests/test_client_request.py]: I'm working on fixing the tests, adding new tests, than I'll come back and remove the ones that don't make sense

Comment (bdraco) [reply tests/test_client_request.py]: They should use the update_body method as we know we can close out the payload.  I'll add notes in the docs.  We can't remove the setter though as it would break backwards compatibility

Comment (bdraco) [reply tests/test_client_request.py]: I think we need to keep it because we have to have backwards compatibility by allowing body to be set.

Comment (Dreamsorcerer) [reply tests/test_client_request.py]: Well, it only became an accessible object 2 days ago. We can always drop it in v4 though.

Comment (bdraco) [reply tests/test_client_request.py]: Yeah I think if we ship this right away, its fine to drop

Comment (bdraco) [reply tests/test_client_request.py]: Let me add a breaking note for that but qualify it as it was added in 3.12.0 so we don't expect anyone is using it after only being released for a few hours

Comment (bdraco) [reply tests/test_client_request.py]: I'm going to make all the `update_*` methods protected as well since someone may start using them otherwise

Comment (bdraco) [reply tests/test_client_request.py]: Well technically they were all accessible before because the request is accessible by making a custom connector

Comment (bdraco) [reply tests/test_client_request.py]: https://github.com/Slzdude/aiohttp-proxy/blob/8810ab6c0daf9f537f2c2bb1ce049fbdae7b9f6e/aiohttp_proxy/connector.py#L97

Comment (bdraco) [reply tests/test_client_request.py]: so I'm going to revert all the changes I just made and keep them accessible

File: aiohttp/client.py [hunk 6]
```diff
@@ -786,6 +793,9 @@ async def _connect_and_send_request(
                         try:
                             redirect_origin = parsed_redirect_url.origin()
                         except ValueError as origin_val_err:
+                            if req._body is not None:
+                                await req._body.close()
+                            resp.close()
```
Comment (bdraco) [inline aiohttp/client.py:798]: This was likely a bug

File: tests/test_client_request.py (hunk 5 above)
Comment (Dreamsorcerer) [reply tests/test_client_request.py]: Yeah, I figured a subclass might overload the methods or something, so wasn't so sure on changing the methods. Making body read-only is probably fine though.

Comment (Dreamsorcerer) [reply tests/test_client_request.py]: > https://github.com/Slzdude/aiohttp-proxy/blob/8810ab6c0daf9f537f2c2bb1ce049fbdae7b9f6e/aiohttp_proxy/connector.py#L97

Oh, so connectors receive the request too, fair enough.

**Comment by bdraco** [comment]
I think the text payloads calculate the wrong size since we want the size in bytes but it returns the size in human readable

File: aiohttp/multipart.py [hunk 7]
```diff
@@ -979,6 +986,11 @@ def size(self) -> Optional[int]:
         return total
 
     def decode(self, encoding: str = "utf-8", errors: str = "strict") -> str:
+        """Return string representation of the multipart data.
+
+        WARNING: This method may do blocking I/O if parts contain file payloads.
+        It should not be called in the event loop. Use as_str() instead.
```
Comment (bdraco) [inline aiohttp/multipart.py]: I removed as_str

File: aiohttp/client.py [hunk 8]
```diff
@@ -734,6 +734,8 @@ async def _connect_and_send_request(
                         redirects += 1
                         history.append(resp)
                         if max_redirects and redirects >= max_redirects:
```
Comment (bdraco) [inline aiohttp/client.py:736]: Almost feels like AsyncExitStack would be better. But too much change

File: aiohttp/client_reqrep.py [hunk 9]
```diff
@@ -190,6 +190,20 @@ class ConnectionKey(NamedTuple):
...
 
+def _warn_if_unclosed_payload(payload: payload.Payload) -> None:
+    """Warn if the payload is not closed.
+
+    Callers must check that the body is a Payload before calling this method.
+    """
+    if not payload.autoclose and not payload.consumed:
+        warnings.warn(
+            "The request body is a payload that needs manual closing. "
+            "It should be closed with await payload.close() to avoid resource leaks.",
```
Comment (bdraco) [inline aiohttp/client_reqrep.py]: Should tell them to use update body here

**Comment by bdraco** [comment]
I'm going to try to get this merged tonight so its not blocking Sam's PRs. I keep finding small nits though

File: aiohttp/client_reqrep.py [hunk 10]
```diff
@@ -654,27 +789,9 @@ async def write_bytes(
...
-                    # Write with length constraint, respecting content_length limit
-                    # If the body is larger than content_length, we truncate it
-                    remaining_bytes = content_length
-                    for chunk in self.body:
-                        await writer.write(chunk[:remaining_bytes])
-                        remaining_bytes -= len(chunk)
-                        if remaining_bytes <= 0:
-                            break
+            if TYPE_CHECKING:
+                assert self._body is not None
```
Comment (bdraco) [inline aiohttp/client_reqrep.py]: There is still a race here because self._body can be set to None during the continue
